nvflare/app_common/tie/connector.py

```diff
@@ -0,0 +1,394 @@
+
+        Returns: None
+
+        """
+        pass
+
+    @abstractmethod
+    def stop(self, fl_ctx: FLContext):
+        """Called by Controller/Executor to stop the connector.
+        If any error occurs, this method should raise an exception.
+
+        Args:
+            fl_ctx: the FL context.
+
+        Returns: None
+
+        """
+        pass
+
+    @abstractmethod
+    def configure(self, config: dict, fl_ctx: FLContext):
+        """Called by Controller/Executor to configure the connector.
+        If any error occurs, this method should raise an exception.
+
+        Args:
+            config: config data
+            fl_ctx: the FL context
+
+        Returns: None
+
+        """
+        pass
+
+    def _is_stopped(self) -> (bool, int):
+        """Called by the connector's monitor to know whether the connector is stopped.
+        Note that this method is not called by Controller/Executor.
+
+        Returns: a tuple of: whether the connector is stopped, and return code (if stopped)
+
+        Note that a non-zero return code is considered abnormal completion of the connector.
+
+        """
+        return self.is_applet_stopped()
+
+    def _monitor(self, fl_ctx: FLContext, connector_stopped_cb):
+        while True:
+            if self.abort_signal.triggered:
+                # asked to abort
+                self.stop(fl_ctx)
+                return
+
+            stopped, rc = self._is_stopped()
+            if stopped:
+                # connector already stopped - notify the caller
+                connector_stopped_cb(rc, fl_ctx)
+                return
+
+            time.sleep(0.1)
+
+    def monitor(self, fl_ctx: FLContext, connector_stopped_cb):
+        """Called by Controller/Executor to monitor the health of the connector.
+
+        The monitor periodically checks the abort signal. Once set, it calls the connector's stop() method
+        to stop the running of the app.
+
+        The monitor also periodically checks whether the connector is already stopped (by calling the is_stopped
+        method). If the connector is stopped, the monitor will call the specified connector_stopped_cb.
+
+        Args:
+            fl_ctx: FL context
+            connector_stopped_cb: the callback function to be called when the connector is stopped.
+
+        Returns: None
+
+        """
+        if not callable(connector_stopped_cb):
+            raise RuntimeError(f"connector_stopped_cb must be callable but got {type(connector_stopped_cb)}")
+
+        # start the monitor in a separate daemon thread!
+        t = threading.Thread(target=self._monitor, args=(fl_ctx, connector_stopped_cb), daemon=True)
+        t.start()
+
+    def start_applet(self, app_ctx: dict, fl_ctx: FLContext):
+        """Start the applet set to the connector.
+        If self.in_process is True, then the applet will be started in a separate thread.
+        If self.in_process is False, then the applet will be started in a separate process.
```

Comment:
The APPLET_ENV_SELF means that the applet itself is responsible for starting the external app (e.g. cli_applet), whereas the in_process flag only applies when the connector is responsible for starting the applet (the XGB model). In this model, the applet runs python directly. In case of APPLET_ENV_SELF, the applet may or may not be python function at all.

Reply:
Got it, I think this "Applet" class is always in Python in our code, but "external app" can be any language as you said.

We can also have APPLET_ENV_PROCESS combine with CLIApplet.
This would mean that Connector will spawn a process to call "self.applet.start()".
Then CLIApplet start will spawn another process to run external script (external app).

For the applet.start() method, whether users want to use spawn process to start external script (CLIApplet) or they just signaling other process to do anything depends on the users' implementation.

Suggest enhance the docstring and code logic to the following:

```
if self.applet_env is Constant.APPLET_ENV_SELF, then "self.applet.start" will be called in the same thread as the connector.
If self.applet_env is Constant.APPLET_ENV_THREAD, then "self.applet.start" will be called in a separate thread.
If self.applet_env is Constant.APPLET_ENV_PROCESS, then "self.applet.start" will be called in a separate process.
```

Another note, I think the Applet class abstract is similar to the Launcher abstract. We might be able to consolidate them later.

Reply:
Changed to use PyApplet for external python code that can be run in_process or not.